h2o-core/src/main/java/hex/DistributionFactory.java

```diff
@@ -439,30 +421,116 @@ public QuantileDistribution(Model.Parameters params){
         super(params);
     }
 
+    @Override
+    public double deviance(double w, double y, double f) { return y > f ? w * quantileAlpha * (y - f) : w * (1 - quantileAlpha) * (f - y); }
+
+    @Override
+    public double negHalfGradient(double y, double f) { return y > f ? 0.5 * quantileAlpha : 0.5 * (quantileAlpha - 1); }
+}
+
+/**
+ * Custom distribution class to customized loss and prediction calculation.
+ * Currently supported only for GBM algorithm.
+ */
+class CustomDistribution extends Distribution {
+    
+    private CustomDistributionWrapper customDistribution;
+    private static CustomDistribution inst;
+    
+    private CustomDistribution(Model.Parameters params){
+        super(params);
+        customDistribution = new CustomDistributionWrapper(CFuncRef.from(params._custom_distribution_func));
+        assert customDistribution != null;
+        assert customDistribution.getFunc() != null;
+        this.setLinkFunction(LinkFunctionFactory.getLinkFunction(customDistribution.getFunc().link()));
```

Comment:
calling overridable methods from a constructor is a bad practice because of descendants of this class can override the method and assume the class is fully constructed when the method is called

Reply:
Yes, sure. That is absolutely true.